In [57]:
import pandas as pd #import the pandas and aliasing as pd
import numpy as np
data = np.array(['a','b','c','d'])
s1 = pd.Series(data)
print(s1)
data = np.array([1,2,3,4])
s2 = pd.Series(data)
print(s2)

0    a
1    b
2    c
3    d
dtype: object
0    1
1    2
2    3
3    4
dtype: int32


In [132]:
data = np.array([ 1, 'b', 3, 'd'])
s3 = pd.Series(data)
print(s3)

0    1
1    b
2    3
3    d
dtype: object


- The datatype of S1 is character

- The datatype of S2 is integer

In [58]:
data = np.array(['a','b','c','d'])
s = pd.Series(data,index=[100,101,102,103])
print(s)


100    a
101    b
102    c
103    d
dtype: object


In [59]:
s = pd.Series([1,2,3,4,5],index = ['a','b','c','d','e'])
print(s[0])

#print 0th positon

1


In [60]:
s = pd.Series([1,2,3,4,5],index = ['a','b','c','d','e'])
print(s[:3])

a    1
b    2
c    3
dtype: int64


In [61]:
print(s[['a']])


a    1
dtype: int64


In [62]:
#retrieve multiple elements
print(s[['a','c','d']])

a    1
c    3
d    4
dtype: int64


In [63]:
import pandas as pd
df = pd.DataFrame()
print(df)

Empty DataFrame
Columns: []
Index: []


In [64]:
data = [1,2,3,4,5]
df1 = pd.DataFrame(data)
print(df1)
data = [['Alex',10],['Bob',12],['Clarke',13]]
df2 = pd.DataFrame(data,columns=['Name','Age'])
print(df2)

   0
0  1
1  2
2  3
3  4
4  5
     Name  Age
0    Alex   10
1     Bob   12
2  Clarke   13


In [65]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],
'Age':[28,34,29,42]}
df = pd.DataFrame(data)
print(df)


    Name  Age
0    Tom   28
1   Jack   34
2  Steve   29
3  Ricky   42


In [21]:
d = {'one' : pd.Series([1, 2, 3], index=['a', 'b', 'c']),
     'two' : pd.Series([1, 2, 3, 4], index=['a', 'b', 'c', 'd'
        ])}
df = pd.DataFrame(d)
print(df)

   one  two
a  1.0    1
b  2.0    2
c  3.0    3
d  NaN    4


- NaN is null

In [67]:
#Table1
data = {'Math' : [90,10,34,23,43],
        'C_programming' : [70,20,25,32,73],
        'Java' : [50,41,67,34,23],
        'Web_programming' : [55,66,23,76,67]
       }
studentDF = pd.DataFrame(data)
print(studentDF)

   Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67


In [69]:
print(studentDF)
print('\n')
print(studentDF[2:4]) #Slice Rows
print('\n')
print(studentDF.loc[0])
print('\n')
print(studentDF.loc[2])
print('\n')
print(studentDF.loc[0:2])

   Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67


   Math  C_programming  Java  Web_programming
2    34             25    67               23
3    23             32    34               76


Math               90
C_programming      70
Java               50
Web_programming    55
Name: 0, dtype: int64


Math               34
C_programming      25
Java               67
Web_programming    23
Name: 2, dtype: int64


   Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23


In [70]:
df = pd.DataFrame([[1, 2], [3, 4]], columns = ['a','b'])
df2 = pd.DataFrame([[5, 6], [7, 8]], columns = ['a','b'])
df = df.append(df2)
print(df)


   a  b
0  1  2
1  3  4
0  5  6
1  7  8


In [103]:
studentDF = {'Math':[90,10,34,23,43],
             'C_programming':[70,20,25,32,73],
             'Java':[50,41,67,34,23],
             'Web_programming':[55,66,23,76,67]
            }
studentDF = pd.DataFrame(studentDF)

studentdf= pd.DataFrame([[59,37,84,82], [41,34,73,55]], columns = ['Math','C_programming','Java','Web_programming'], index = [5,6])

studentDF = studentDF.append(studentdf)
print(studentDF)

   Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82
6    41             34    73               55


In [100]:
# Drop rows with label 0
studentDF = studentDF.drop(6)
print(studentDF)


   Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82


In [101]:
data = {'Name':['Tom', 'Jack', 'Steve', 'Ricky'],
        'Age':[28,34,29,42]
        }
df = pd.DataFrame(data)
print(df.sum())

Name    TomJackSteveRicky
Age                   133
dtype: object


In [107]:
studentDF = {'Math':[90,10,34,23,43],
             'C_programming':[70,20,25,32,73],
             'Java':[50,41,67,34,23],
             'Web_programming':[55,66,23,76,67]
            }
studentDF = pd.DataFrame(studentDF)

studentdf= pd.DataFrame([[59,37,84,82], [41,34,73,55]], columns = ['Math','C_programming','Java','Web_programming'], index = [5,6])

studentDF = studentDF.append(studentdf)
studentDF.sum()

Math               300
C_programming      291
Java               372
Web_programming    424
dtype: int64

In [108]:
studentDF.sum(axis = 1)

0    265
1    137
2    149
3    165
4    206
5    262
6    203
dtype: int64

In [109]:
studentDF.count()

Math               7
C_programming      7
Java               7
Web_programming    7
dtype: int64

In [110]:
studentDF.mean()

Math               42.857143
C_programming      41.571429
Java               53.142857
Web_programming    60.571429
dtype: float64

In [111]:
studentDF.median()

Math               41.0
C_programming      34.0
Java               50.0
Web_programming    66.0
dtype: float64

In [112]:
studentDF.mode()

Math  C_programming  Java  Web_programming
0    10             20    23             55.0
1    23             25    34              NaN
2    34             32    41              NaN
3    41             34    50              NaN
4    43             37    67              NaN
5    59             70    73              NaN
6    90             73    84              NaN

In [113]:
studentDF.std()

Math               25.944997
C_programming      21.235639
Java               22.251806
Web_programming    19.329228
dtype: float64

In [114]:
studentDF.min()

Math               10
C_programming      20
Java               23
Web_programming    23
dtype: int64

In [115]:
studentDF.max()

Math               90
C_programming      73
Java               84
Web_programming    82
dtype: int64

In [116]:
studentDF.abs()

Math  C_programming  Java  Web_programming
0    90             70    50               55
1    10             20    41               66
2    34             25    67               23
3    23             32    34               76
4    43             73    23               67
5    59             37    84               82
6    41             34    73               55

In [117]:
studentDF.prod()

Math                 73207164600
C_programming       102854080000
Java                658624016400
Web_programming    1917341170800
dtype: int64

In [118]:
studentDF.cumsum()

Math  C_programming  Java  Web_programming
0    90             70    50               55
1   100             90    91              121
2   134            115   158              144
3   157            147   192              220
4   200            220   215              287
5   259            257   299              369
6   300            291   372              424

In [119]:
studentDF.cumprod()

Math  C_programming          Java  Web_programming
0           90             70            50               55
1          900           1400          2050             3630
2        30600          35000        137350            83490
3       703800        1120000       4669900          6345240
4     30263400       81760000     107407700        425131080
5   1785540600     3025120000    9022246800      34860748560
6  73207164600   102854080000  658624016400    1917341170800

In [120]:
import pandas as pd
import numpy as np
#Create a Dictionary of series
d = {'Name':pd.Series(['Tom','James','Ricky','Vin','Steve','Smith','Jack','Lee','David','Gasper','Betina','Andres']),
     'Age':pd.Series([25,26,25,23,30,29,23,34,40,30,51,46]),
     'Rating':pd.Series([4.23,3.24,3.98,2.56,3.20,4.6,3.8,3.78,2.98,4.80,4.10,3.65])
    }
#Create a DataFrame
df = pd.DataFrame(d)
df.describe()

Age     Rating
count  12.000000  12.000000
mean   31.833333   3.743333
std     9.232682   0.661628
min    23.000000   2.560000
25%    25.000000   3.230000
50%    29.500000   3.790000
75%    35.500000   4.132500
max    51.000000   4.800000

In [121]:
studentDF.describe()

Math  C_programming       Java  Web_programming
count   7.000000       7.000000   7.000000         7.000000
mean   42.857143      41.571429  53.142857        60.571429
std    25.944997      21.235639  22.251806        19.329228
min    10.000000      20.000000  23.000000        23.000000
25%    28.500000      28.500000  37.500000        55.000000
50%    41.000000      34.000000  50.000000        66.000000
75%    51.000000      53.500000  70.000000        71.500000
max    90.000000      73.000000  84.000000        82.000000

- The mean value of Math score is 42.857143
- The score at the third quartile of Java is 70.000000
- The standard deviation of Web programming score is 19.329228

In [122]:
studentDF = studentDF.reindex([0,1,2,3,4,5,6,7])
print(studentDF)


   Math  C_programming  Java  Web_programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6  41.0           34.0  73.0             55.0
7   NaN            NaN   NaN              NaN


- There are 4 NaN appear in the table studentDF.

In [124]:
df = pd.DataFrame(np.random.randn(5, 3), index=['a', 'c', 'e', 'f', 'h'],columns=['one', 'two', 'three'])
df = df.reindex(['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h'])
print(df['one'].isnull())
print(df['two'].notnull())

a    False
b     True
c    False
d     True
e    False
f    False
g     True
h    False
Name: one, dtype: bool
a     True
b    False
c     True
d    False
e     True
f     True
g    False
h     True
Name: two, dtype: bool


In [125]:
print(studentDF['Math'].isnull())
print(studentDF['C_programming'].isnull())
print(studentDF['Java'].isnull())
print(studentDF['Web_programming'].isnull())

0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
Name: Math, dtype: bool
0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
Name: C_programming, dtype: bool
0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
Name: Java, dtype: bool
0    False
1    False
2    False
3    False
4    False
5    False
6    False
7     True
Name: Web_programming, dtype: bool


In [126]:
s1 = studentDF
s2 = studentDF
s1 = s1.fillna(0)
s2 = s2.fillna(method='pad')
print(s1)
print(s2)

   Math  C_programming  Java  Web_programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6  41.0           34.0  73.0             55.0
7   0.0            0.0   0.0              0.0
   Math  C_programming  Java  Web_programming
0  90.0           70.0  50.0             55.0
1  10.0           20.0  41.0             66.0
2  34.0           25.0  67.0             23.0
3  23.0           32.0  34.0             76.0
4  43.0           73.0  23.0             67.0
5  59.0           37.0  84.0             82.0
6  41.0           34.0  73.0             55.0
7  41.0           34.0  73.0             55.0


- NaN position, the value of s1 is 0.0. but the value of s2 is the same as previous row in the same column.

In [127]:
left = pd.DataFrame({'id':[1,2,3,4,5],
                     'Name': ['Alex', 'Amy', 'Allen', 'Alice', 'Ayoung'],
                     'subject_id':['sub1','sub2','sub4','sub6','sub5']
                    })
right = pd.DataFrame({'id':[1,2,3,4,5],
                      'subject_name':['Math','English','Social Science','Programming','Physics']
                    })
print(left)
print(right)


   id    Name subject_id
0   1    Alex       sub1
1   2     Amy       sub2
2   3   Allen       sub4
3   4   Alice       sub6
4   5  Ayoung       sub5
   id    subject_name
0   1            Math
1   2         English
2   3  Social Science
3   4     Programming
4   5         Physics


In [128]:
merged_table = pd.merge(left,right,on='id')
print(merged_table)

   id    Name subject_id    subject_name
0   1    Alex       sub1            Math
1   2     Amy       sub2         English
2   3   Allen       sub4  Social Science
3   4   Alice       sub6     Programming
4   5  Ayoung       sub5         Physics


In [129]:
section = pd.DataFrame({'id':[0,1,2,3,4,5],'section': ['one', 'one', 'one', 'two', 'two', 'two']})

student = pd.DataFrame({'id':[0,1,2,3,4,5],'Name': ['Mat', 'Mike', 'John', 'Peecha', 'Prabu', 'Bunchu'],
                        'math':[34,12,34,12,65,23],
                        'english':[23,13,51,62,12,42]
                       })

score = pd.merge(section,student,on='id')
print(score)

   id section    Name  math  english
0   0     one     Mat    34       23
1   1     one    Mike    12       13
2   2     one    John    34       51
3   3     two  Peecha    12       62
4   4     two   Prabu    65       12
5   5     two  Bunchu    23       42


In [131]:
score_sec = score.groupby('section')                      
score_sec.sum()

id  math  english
section                   
one       3    80       87
two      12   100      116